sw/device/lib/rv_timer.c

```diff
@@ -4,42 +4,67 @@
 
 #include "sw/device/lib/rv_timer.h"
 
-#include "rv_timer_regs.h"  // Generated.
-#include "sw/device/lib/common.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_rv_timer.h"
 #include "sw/device/lib/runtime/ibex.h"
 
 #define RV_TIMER0_BASE_ADDR 0x40080000
 #define HART_CFG_ADDR_GAP 0x100
 
-static const uint32_t NS_IN_S = 1000 * 1000 * 1000;
+// Timer for hart #0.
+static dif_rv_timer_t timer0;
 
 void rv_timer_set_us_tick(uint32_t hart) {
-  uint32_t ticks_per_us = (uint32_t)((1000 * kIbexClockFreqHz) / NS_IN_S) - 1;
+  if (hart != 0) {
+    return;
+  }
 
-  REG32(RV_TIMER_CFG0(0) + hart * 4) =
-      (ticks_per_us & RV_TIMER_CFG0_PRESCALE_MASK) |
-      (1 << RV_TIMER_CFG0_STEP_OFFSET);
+  dif_rv_timer_config_t config = {
+      .base_addr = mmio_region_from_addr(RV_TIMER0_BASE_ADDR),
+      .hart = 0,
+  };
+  (void)dif_rv_timer_approximate_tick_params(kIbexClockFreqHz, 1000 * 1000,
+                                             &config.tick_params);
+
+  (void)dif_rv_timer_init(config, &timer0);
+}
+
+uint64_t rv_timer_get(uint32_t hart) {
+  if (hart != 0) {
+    return 0ULL - 1;
+  }
+
+  uint64_t value;
+  (void)dif_rv_timer_counter_read(&timer0, &value);
+  return value;
 }
 
 void rv_timer_set_cmp(uint32_t hart, uint64_t cmp) {
-  REG32(RV_TIMER_COMPARE_UPPER0_0(0) + hart * HART_CFG_ADDR_GAP) = -1;
-  REG32(RV_TIMER_COMPARE_LOWER0_0(0) + hart * HART_CFG_ADDR_GAP) =
-      (uint32_t)cmp;
-  REG32(RV_TIMER_COMPARE_UPPER0_0(0) + hart * HART_CFG_ADDR_GAP) =
-      (uint32_t)(cmp >> 32);
+  if (hart != 0) {
+    return;
+  }
+
+  (void)dif_rv_timer_arm(&timer0, cmp);
 }
 
 void rv_timer_ctrl(uint32_t hart, bool en) {
-  REG32(RV_TIMER_CTRL(0)) = (en) ? SETBIT(REG32(RV_TIMER_CTRL(0)), hart)
-                                 : CLRBIT(REG32(RV_TIMER_CTRL(0)), hart);
+  if (hart != 0) {
+    return;
+  }
+
+  (void)dif_rv_timer_set_enabled(&timer0,
+                                 en ? kDifRvTimerEnabled : kDifRvTimerDisabled);
 }
 
 void rv_timer_intr_enable(uint32_t hart, bool en) {
-  REG32(RV_TIMER_INTR_ENABLE0(0)) =
-      (en) ? SETBIT(REG32(RV_TIMER_INTR_ENABLE0(0)), hart)
-           : CLRBIT(REG32(RV_TIMER_INTR_ENABLE0(0)), hart);
+  if (hart != 0) {
+    return;
+  }
+
+  (void)dif_rv_timer_set_irq_enabled(
+      &timer0, en ? kDifRvTimerEnabled : kDifRvTimerDisabled);
 }
 
 void rv_timer_clr_all_intrs(void) {
```

Comment:
For the sake of consistency, this function should have a parameter for hart index like the rest of the functions in this file.

Reply:
Should it? I think clear_all affects all harts simultaneously.

Reply:
I thought this was omitted because we only have a single timer now and interpreted `clear_all` as all interrupts for a given hart.

Reply:
You're right. I'm just going to get rid of rv_timer.h/c altogether, since it's only used in the timer test.